File: Ackermann/Lua/Yonaba/ackermann.lua
-- Computes values of Ackermann function
-- See: http://en.wikipedia.org/wiki/Ackermann_function

-- m : a nonnegative integer
-- n : a nonnegative integer
-- return : the value of Ackermann(m, n)
local function ackermann(m, n)
  if m == 0 then return n + 1 end
  if n == 0 then return ackermann(m-1,1) end
  return ackermann(m-1, ackermann(m, n-1))
end

return ackermann

File: Ackermann/C++/alsuga/ackermann.cc
long long ackermann(long long m, long long n) {
  if(m == 0)
    return n + 1;
  if(n == 0)
    return ackermann(m - 1, 1);
  return ackermann(m - 1, ackermann(m , n - 1));
}
